temporal-testing/src/main/java/io/temporal/testing/IdempotentTimeLocker.java

```diff
@@ -0,0 +1,70 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.testing;
+
+import io.grpc.Context;
+import io.temporal.api.testservice.v1.LockTimeSkippingRequest;
+import io.temporal.api.testservice.v1.UnlockTimeSkippingRequest;
+import io.temporal.serviceclient.TestServiceStubs;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Used to ensure that multiple TimeLockingWorkflowStubs that are blocked at the same time from
+ * multiple threads execute unlock only once and the lock only once.
+ */
+class IdempotentTimeLocker {
+  private final TestServiceStubs testServiceStubs;
+  private final AtomicInteger count = new AtomicInteger(1);
+
+  IdempotentTimeLocker(TestServiceStubs testServiceStubs) {
+    this.testServiceStubs = testServiceStubs;
+  }
+
+  public void lockTimeSkipping(String caller) {
```

Comment:
There may be a problem with this approach - when a given call to `lockTimeSkipping` returns, it is _not_ guaranteed that time skipping has been locked. Any call subsequent to the first call might return early _while that first call is still executing the RPC call_. This leaves a window where this method has returned but time skipping is not (yet) locked.

Reply:
This code is a refactoring of the existing one, not a new code, so I'm not gonna tackle this in this PR.
But this (return without lock/unlock actually happening) actually shouldn't be an issue for the caller code. The only thing this wrapper is checking is that one WorkflowStub doesn't cause several locks/unlocks.
There is another problem with this code though: https://github.com/temporalio/sdk-java/issues/392 But again, not something I'm gonna tackle here.